doc/Configuration.rst
****************************
Configuration and management
****************************

DNS Probe is configured at startup through a `YAML <https://yaml.org/>`_ file. Runtime remote management API is planned
to be implemented in future releases. When everything is in place, the API is expected to provide the following functionality:

* configuration datastores and data-driven API for configuration and management

* remote configuration and management via standard protocols

* real-time access to state data and statistics

* RPC operations

* event notifications

* unified configuration and management that includes other system components such as DNS servers or routing daemons.

Data model
==========

DNS Probe uses local file in YAML format to load configuration at startup. Its schema tree looks as follows::

   +--rw <instance-id>?
      +--rw coremask? <uint64>
      +--rw dnstap-socket-group? <string>
      +--rw dnstap-socket-list?* <string>
      +--rw dns-ports?* <uint16>
      +--rw export
      |  +--rw asn-maxmind-db? <string>
      |  +--rw backup-remote-ip-address? <string>
      |  +--rw backup-remote-port? <uint16>
      |  +--rw cdns-blocks-per-file? <uint64>
      |  +--rw cdns-export-response-rr? <boolean>
      |  +--rw cdns-fields?* <string>
      |  +--rw cdns-records-per-block? <uint64>
      |  +--rw country-maxmind-db? <string>
      |  +--rw export-dir? <string>
      |  +--rw export-format? <enumeration>
      |  +--rw file-compression? <boolean>
      |  +--rw file-name-prefix? <string>
      |  +--rw file-size-limit? <uint64>
      |  +--rw kafka-address-family? <enumeration>
      |  +--rw kafka-brokers? <string>
      |  +--rw kafka-ca-location? <string>
      |  +--rw kafka-cert-location? <string>
      |  +--rw kafka-key-location? <string>
      |  +--rw kafka-key-password? <string>
      |  +--rw kafka-partition? <string>
      |  +--rw kafka-security-protocol? <enumeration>
      |  +--rw kafka-sasl-mechanism? <enumeration>
      |  +--rw kafka-sasl-password? <string>
      |  +--rw kafka-sasl-username? <string>
      |  +--rw kafka-topic? <string>
      |  +--rw location? <enumeration>
      |  +--rw parquet-records-per-file? <uint64>
      |  +--rw pcap-export? <enumeration>
      |  +--rw remote-ca-cert? <string>
      |  +--rw remote-ip-address? <string>
      |  +--rw remote-port? <uint16>
      |  +--rw timeout? <uint32>
      +--rw interface-list?* <string>
      +--rw ipv4-allowlist?* <string>
      +--rw ipv4-denylist?* <string>
      +--rw ipv6-allowlist?* <string>
      +--rw ipv6-denylist?* <string>
      +--rw ip-anonymization
      |  +--rw anonymize-ip? <boolean>
      |  +--rw encryption? <enumeration>
      |  +--rw key-path? <string>
      +--rw knot-socket-count? <uint32>
      +--rw knot-socket-path? <string>
      +--rw log-file? <string>
      +--rw pcap-list?* <string>
      +--rw raw-pcap? <boolean>
      +--rw statistics
      |  +--rw backup-remote-ip? <string>
      |  +--rw backup_remote-port? <uint16>
      |  +--rw export-dir? <string>
      |  +--rw export-stats? <boolean>
      |  +--rw kafka-address-family? <enumeration>
      |  +--rw kafka-brokers? <string>
      |  +--rw kafka-ca-location? <string>
      |  +--rw kafka-cert-location? <string>
      |  +--rw kafka-key-location? <string>
      |  +--rw kafka-key-password? <string>
      |  +--rw kafka-partition? <string>
      |  +--rw kafka-security-protocol? <enumeration>
      |  +--rw kafka-sasl-mechanism? <enumeration>
      |  +--rw kafka-sasl-password? <string>
      |  +--rw kafka-sasl-username? <string>
      |  +--rw kafka-topic? <string>
      |  +--rw location? <enumeration>
      |  +--rw moving-avg-window? <uint16>
      |  +--rw remote-ca-cert? <string>
      |  +--rw remote-ip? <string>
      |  +--rw remote-port? <uint16>
      |  +--rw stats-fields?* <string>
      |  +--rw stats-per-ip <boolean>
      |  +--rw stats-timeout? <uint32>
      +--rw tcp-table
      |  +--rw concurrent-connections? <uint32>
      |  +--rw timeout? <uint64>
      +--rw transaction-table
         +--rw match-qname? <boolean>
         +--rw max-transactions? <uint32>
         +--rw query-timeout? <uint64>

Configuring DNS Probe via YAML file
===================================

.. Note:: Configuration interfaces are somewhat spartan and rudimentary in the current version of DNS Probe. More user-friendly approaches are being worked on.

After installation, a default YAML configuration file is installed at *<INSTALL_DIR>/etc/dns-probe-<BACKEND>/dns-probe.yml*.
By default, DNS Probe will try to load configuration from this file at startup.

The contents of the default file can be edited by user or a different YAML configuration file can be provided to DNS Probe
via the `-c` command line parameter.

In future releases it will be possible to configure and manage DNS Probe remotely using a remote management API.

Configuration parameters
========================

All YAML configuration parameters appear in the top-level ``instance-id`` container. ``instance-id`` uniquely
identifies given instance of DNS Probe with its configuration. Instance ID of DNS Probe instance can be set at startup
by the `-n` command line parameter. Only configuration with this particular ``instance-id`` will then be loaded
from YAML configuration file. User can configure multiple instances of DNS Probe from one YAML file like this.

A special ``instance-id`` is *default*. Configuration set under *default* instance is loaded by all instances
of DNS Probe before its specific instance configuration. A common configuration for all instances can be set
using this special ``instance-id``.

DNS Probe binaries internally contain default values for all configuration options so the *default* instance
configuration can be ommited from YAML configuration file. This means that user only has to specify options
differing from default values for a specific instance of DNS Probe in the YAML file.

For more information about the YAML configuration file format see :doc:`Default YAML file <YAMLfile>`.

List of configuration parameters
--------------------------------

This section lists all configuration parameters in alphabetical order.

anonymize-ip
^^^^^^^^^^^^

:data node: ``<instance-id>/ip-anonymization/anonymize-ip``
:default: **false**

If this flag is true, client IP addresses in exported data (Parquet or C-DNS, NOT optional PCAPs) will be anonymized using Crypto-PAn prefix-preserving algorithm.

asn-maxmind-db
^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/asn-maxmind-db``
:default: empty

Path to Maxmind ASN database. If this option is set to a valid database file, the ``asn`` implementation field in exported Parquets or C-DNS will be filled with Autonomous System Number (ASN) based on client's IP address.

.. _backup-remote-ip-address:

backup-remote-ip-address
^^^^^^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/backup-remote-ip-address``
:default: empty

Backup IP address for remote export of the DNS records. Will be used if :ref:`location` is set to ``remote``.

.. _backup-stats-remote-ip:

backup-remote-ip
^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/statistics/backup-remote-ip``
:default: empty

Backup IP address for remote export of run-time statistics. Will be used if :ref:`stats-location` is set to ``remote``
and :ref:`export-stats` is set to **true**.

.. _backup-remote-port:

backup-remote-port
^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/backup-remote-port``
:default: 6378

Backup tranport protocol port number for remote export of the DNS records. Will be used if :ref:`location` is set to ``remote``.

.. _backup-stats-remote-port:

backup-remote-port
^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/statistics/backup-remote-port``
:default: 6379

Backup transport protocol port number for remote export of run-time statistics. Will be used if :ref:`stats-location`
is set to ``remote`` and :ref:`export-stats` is set to **true**.

.. _cdns-blocks-per-file:

cdns-blocks-per-file
^^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/cdns-blocks-per-file``
:default: 0

This parameter takes effect only if ``cdns`` is set in :ref:`export-format`. It specifies the maximum number of C-DNS blocks written to one exported file (see `Section 7.3.2 <https://tools.ietf.org/html/rfc8618#section-7.3.2>`_ in [RFC8618]_). If this limit is reached, the export file is closed and a new one started.

The default value of 0 means that there is no limit.

.. _cdns-export-response-rr:

cdns-export-response-rr
^^^^^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/cdns-export-response-rr``
:default: **false**

If this flag is set to **true**, exported C-DNS files will contain full Answer and Additional RRs from responses in each record.

**NOTE:** Won't work for traffic captured via Knot interface as this data doesn't contain full RRs.

cdns-fields
^^^^^^^^^^^

:data node: ``<instance-id>/export/cdns-fields``
:default: all fields

This parameter takes effect only if ``cdns`` is set in :ref:`export-format`. It is a bit set that determines which fields from the C-DNS schema defined in [RFC8618]_ will be included in the exported transaction records.

.. _cdns-records-per-block:

cdns-records-per-block
^^^^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/cdns-records-per-block``
:default: 10000

This parameter takes effect only if ``cdns`` is set in :ref:`export-format`. It specifies the maximum number of exported DNS transaction records per one C-DNS block, see `Section 7.3.2 <https://tools.ietf.org/html/rfc8618#section-7.3.2>`_ in [RFC8618]_.

The default value of 10000 corresponds to the recommendation in `Appendix C.6 <https://tools.ietf.org/html/rfc8618#appendix-C.6>`_ of [RFC8618]_.

concurrent-connections
^^^^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/tcp-table/concurrent-connections``
:default: 131072

The value of this parameter must be a power of 2. It specifies the maximum number of TCP connections that DNS Probe can handle at any given time, which in turn affects the size of in-memory data structures allocated for keeping the status of TCP connections.

The default value of 131072 (2^17) was determined experimentally – it takes into account the default value for :ref:`max-transactions` and the current common ratio of DNS traffic over UDP and TCP. It is recommended to adjust this parameter to actual traffic circumstances in order to optimize memory consumption.

coremask
^^^^^^^^

:data node: ``<instance-id>/coremask``
:default: 7

Bitmask indicating which CPU cores should DNS Probe use. At least 3 CPU cores are needed, see :ref:`dns-probe-arch`. Setting more than 3 cores in the bitmask will spawn more worker threads that are used for processing incoming packets.

The default value of 7 indicates that DNS Probe should use the first 3 CPU cores with IDs of 0, 1 and 2.

country-maxmind-db
^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/country-maxmind-db``
:default: empty

Path to Maxmind Country database. If this option is set to a valid database file, the ``country`` field in exported Parquets or ``country-code`` implementation field in exported C-DNS will be filled with ISO 3166-1 country code based on client's IP address.

dnstap-socket-group
^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/dnstap-socket-group``
:default: empty

Name of existing user group under which to create dnstap sockets specified in :ref:`dnstap-socket-list`. By default the group of probe's process is used.

.. _dnstap-socket-list:

dnstap-socket-list
^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/dnstap-socket-list``
:default: empty

List of unix sockets to process dnstap data from in addition to sockets passed with '-d'
command line parameter.

.. _dns-ports:

dns-ports
^^^^^^^^^

:data node: ``<instance-id>/dns-ports``
:default: 53

List of transport protocol port numbers that DNS Probe will check for in
incoming packets to recognize DNS traffic.

The default value of 53 is the standard DNS server port as defined
in [RFC1035]_.

.. _encryption:

encryption
^^^^^^^^^^

:data node: ``<instance-id>/ip-anonymization/encryption``
:default: ``aes``

Encryption algorithm to be used during anonymization of client IP addresses if enabled. Four options currently supported:

``aes``
   AES encryption algorithm.

``blowfish``
   Blowfish encryption algorithm.

``md5``
   MD5 hash function.

``sha1``
   SHA1 hash function.

.. _export-dir:

export-dir
^^^^^^^^^^

:data node: ``<instance-id>/export/export-dir``
:default: ``.``

Path to an existing local directory for storing export files.

The default value of ``.`` means that DNS Probe will use the current working directory from which it was launched.

.. _stats-export-dir:

export-dir
^^^^^^^^^^

:data node: ``<instance-id>/statistics/export-dir``
:default: ``.``

Path to an existing local directory for storing run-time statistics in JSON.

The default value of ``.`` means that DNS Probe will use the current working directory from which it was launched.

.. _export-format:

export-format
^^^^^^^^^^^^^

:data node: ``<instance-id>/export/export-format``
:default: ``parquet``

This value indicates the format for exporting records about
DNS transactions. Three options are currently supported:

``parquet``
   `Apache Parquet <https://parquet.apache.org/>`_ columnar format

``cdns``
   Compacted-DNS (C-DNS) [RFC8618]_.

``json``
   JSON [RFC8259]_

.. _export-stats:

export-stats
^^^^^^^^^^^^

:data node: ``<instance-id>/statistics/export-stats``
:default: **false**

If this flag is true, run-time statistics will be exported in JSON format every :ref:`stats-timeout` seconds.

file-compression
^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/file-compression``
:default: **true**

If this flag is true, the exported Parquet or C-DNS files will be
compressed with GZIP. C-DNS export files are compressed in their
entirety, and suffix ``.gz`` is appended to their names. Parquet
format implementation used by DNS Probe compresses only selected parts
of the file, and there is no ``.gz``.

.. _file-name-prefix:

file-name-prefix
^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/file-name-prefix``
:default: ``dns_``

This option represents the prefix that is prepended to the name of all
files exported by DNS Probe.

file-size-limit
^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/file-size-limit``
:default: 0

This parameter specifies the maximum size of export file in megabytes. It is currently used only for rotating files of the auxiliary PCAP export described in :ref:`pcap-export` below, because estimating the size of data in Parquet or C-DNS files is quite tricky if not impossible.

The default value of 0 means that the export file will never be closed just based on its size.

.. _interface-list:

interface-list
^^^^^^^^^^^^^^

:data node: ``<instance-id>/interface-list``
:default: empty

List of network interfaces to process traffic from in addition to interfaces passed with `-i`
command line parameter.

Fill either with NIC interface names such as `eth0` or alternatively with PCI IDs when using DPDK backend
and binding NICs to DPDK-compatible drivers manually.

.. _ipv4-allowlist:

ipv4-allowlist
^^^^^^^^^^^^^^

:data node: ``<instance-id>/ipv4-allowlist``
:default: empty

List of allowed IPv4 addresses and prefixes to process traffic from.

By default all IPv4 addressess are allowed.

ipv4-denylist
^^^^^^^^^^^^^

:data node: ``<instance-id>/ipv4-denylist``
:default: empty

List of IPv4 addresses and prefixes from which to NOT process traffic.

By default all IPv4 addresses are allowed.

If :ref:`ipv4-allowlist` is not empty this configuration item doesn't have any effect.

.. _ipv6-allowlist:

ipv6-allowlist
^^^^^^^^^^^^^^

:data node: ``<instance-id>/ipv6-allowlist``
:default: empty

List of allowed IPv6 addresses and prefixes to process traffic from.

By default all IPv6 addresses are allowed.

ipv6-denylist
^^^^^^^^^^^^^

:data node: ``<instance-id>/ipv6-denylist``
:default: empty

List of IPv6 addresses and prefixes from which to NOT process traffic.

By default all IPv6 addresses are allowed.

If :ref:`ipv6-allowlist` is not empty this configuration item doesn't have any effect.

kafka-address-family
^^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/kafka-address-family``
:default: ``any``

Force IP address family for connection to Kafka brokers for export of DNS records

Valid values are ``any``, ``v4``, ``v6``.

kafka-address-family
^^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/statistics/kafka-address-family``
:default: ``any``

Force IP address family for connection to Kafka brokers for export of run-time statistics.

Valid values are ``any``, ``v4``, ``v6``.

.. _kafka-brokers:

kafka-brokers
^^^^^^^^^^^^^

:data node: ``<instance-id>/export/kafka-brokers``
:default: ``127.0.0.1``

Comma separated list of Kafka brokers (host or host:port) to export DNS records to.
Will be used if :ref:`location` is set to ``kafka``.

.. _stats-kafka-brokers:

kafka-brokers
^^^^^^^^^^^^^

:data node: ``<instance-id>/statistics/kafka-brokers``
:default: ``127.0.0.1``

Comma separated list of Kafka brokers (host or host:port) to export run-time statistics to.
Will be used if :ref:`stats-location` is set to ``kafka``.

kafka-ca-location
^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/kafka-ca-location``
:default: empty

File or directory path to CA certificates(s) for verifying Kafka broker's key.

By default OpenSSL's default directory with CA certificates is used.

kafka-ca-location
^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/statistics/kafka-ca-location``
:default: empty

File or directory path to CA certificates(s) for verifying Kafka broker's key.

By default OpenSSL's default directory with CA certificates is used.

kafka-cert-location
^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/kafka-cert-location``
:default: empty

Path (including file's name) to public key (PEM) used for authentication to Kafka cluster
when :ref:`kafka-security-protocol` is set to ``ssl`` or ``sasl_ssl``.

kafka-cert-location
^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/statistics/kafka-cert-location``
:default: empty

Path (including file's name) to public key (PEM) used for authentication to Kafka cluster
when :ref:`stats-kafka-security-protocol` is set to ``ssl`` or ``sasl_ssl``.

.. _kafka-key-location:

kafka-key-location
^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/kafka-key-location``
:default: empty

Path (including file's name) to private key (PEM) used for authentication to Kafka cluster
when :ref:`kafka-security-protocol` is set to ``ssl`` or ``sasl_ssl``.

.. _stats-kafka-key-location:

kafka-key-location
^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/statistics/kafka-key-location``
:default: empty

Path (including file's name) to private key (PEM) used for authentication to Kafka cluster
when :ref:`stats-kafka-security-protocol` is set to ``ssl`` or ``sasl_ssl``.

kafka-key-password
^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/kafka-key-password``
:default: empty

Private key passphrase for key set in :ref:`kafka-key-location`.

kafka-key-password
^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/statistics/kafka-key-password``
:default: empty

Private key passphrase for key set in :ref:`stats-kafka-key-location`.

.. _kafka-partition:

kafka-partition
^^^^^^^^^^^^^^^

:data-node: ``<instance-id>/export/kafka-partition``
:default: empty

Kafka message key that will be used to assign all messages (DNS records) to specific partition within Kafka topic.

.. _stats-kafka-partition:

kafka-partition
^^^^^^^^^^^^^^^

:data-node: ``<instance-id>/statistics/kafka-partition``
:default: empty

Kafka message key that will be used to assign all messages (run-time statistics) to specific partition within Kafka topic.

.. _kafka-security-protocol:

kafka-security-protocol
^^^^^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/kafka-security-protocol``
:default: ``plaintext``

Security protocol used to communicate with Kafka brokers.

Valid values are ``plaintext``, ``ssl``, ``sasl_plaintext`` and ``sasl_ssl``.

.. _stats-kafka-security-protocol:

kafka-security-protocol
^^^^^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/statistics/kafka-security-protocol``
:default: ``plaintext``

Security protocol used to communicate with Kafka brokers.

Valid values are ``plaintext``, ``ssl``, ``sasl_plaintext`` and ``sasl_ssl``.

.. _kafka-sasl-mechanism:

kafka-sasl-mechanism
^^^^^^^^^^^^^^^^^^^^

:data node: ``<instanceid>/export/kafka-sasl-mechanism``
:default: ``plain``

SASL mechanism to use for authentication to Kafka brokers.

Valid values are ``plain``, ``scram-sha-256`` and ``scram-sha-512``.

.. _stats-kafka-sasl-mechanism:

kafka-sasl-mechanism
^^^^^^^^^^^^^^^^^^^^

:data node: ``<instanceid>/statistics/kafka-sasl-mechanism``
:default: ``plain``

SASL mechanism to use for authentication to Kafka brokers.

Valid values are ``plain``, ``scram-sha-256`` and ``scram-sha-512``.

kafka-sasl-password
^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/export/kafka-sasl-password``
:default: empty

Password for SASL authentication to Kafka brokers.

kafka-sasl-password
^^^^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/statistics/kafka-sasl-password``
:default: empty

Password for SASL authentication to Kafka brokers.

kafka-sasl-username
^^^^^^^^^^^^^^^^^^^

:data node: ``<instanceid>/export/kafka-sasl-username``
:default: empty

Username for SASL authentication to Kafka brokers.

kafka-sasl-username
^^^^^^^^^^^^^^^^^^^

:data node: ``<instanceid>/statistics/kafka-sasl-username``
:default: empty

Username for SASL authentication to Kafka brokers.

.. _kafka-topic:

kafka-topic
^^^^^^^^^^^

:data-node: ``<instance-id>/export/kafka-topic``
:default: ``dns-probe``

Kafka topic to export DNS records to. Will be used if :ref:`location` is set to ``kafka``.

.. _stats-kafka-topic:

kafka-topic
^^^^^^^^^^^

:data-node: ``<instance-id>/statistics/kafka-topic``
:default: ``dns-probe-stats``

Kafka topic to export run-time statistics to. Will be used if :ref:`stats-location` is set to ``kafka``.

key-path
^^^^^^^^

:data node: ``<instance-id>/ip-anonymization/key-path``
:default: ``key.cryptopant``

Path (including file's names) to the file with encryption key that is to be used for client IP anonymization if enabled.
If the file doesn't exist, it is generated by the probe.

The key needs to be compatible with the encryption algorithm set in the :ref:`encryption` option. User should generate
the key using `scramble_ips` tool installed by the cryptopANT dependency like this:

.. code:: shell

   scramble_ips --newkey --type=<encryption> <key_file>

knot-socket-count
^^^^^^^^^^^^^^^^^

:data-node: ``<instance-id>/knot-socket-count``
:default: ``0``

Number of Knot interface sockets to create in :ref:`knot-socket-path` directory.
Might get overriden by `-k` comand line parameter.

.. _knot-socket-path:

knot-socket-path
^^^^^^^^^^^^^^^^

:data-node: ``<instance-id>/knot-socket-path``
:default: ``/tmp``

Path to directory in which to create unix sockets for reading Knot interface data.
Might get overriden by `-s` command line parameter.

.. _location:

location
^^^^^^^^

:data node: ``<instance-id>/export/location``
:default: ``local``

Location for the storage of exported DNS records. Determines if data is stored to local file or sent
to remote server.

Valid values are ``local`` and ``remote``.

.. _stats-location:

location
^^^^^^^^

:data node: ``<instance-id>/statistics/location``
:default: ``local``

Location for the storage of exported run-time statistics in JSON. Determines if data is stored to
local file or sent to remote server.

Valid values are ``local`` and ``remote``.

log-file
^^^^^^^^

:data node: ``<instance-id>/log-file``
:default: empty

Path (including file's name) to log file for storing probe's logs (e.g. `/var/log/dns-probe.log`).
Might get overriden by `-l` command line parameter.

By default logs are written to `stdout`.

match-qname
^^^^^^^^^^^

:data node: ``<instance-id>/transaction-table/match-qname``
:default: **false**

By default, the 5-tuple of source and destination IP address, source and destination port, and transport protocol is used to match a DNS query with the corresponding response. If this parameter is set to **true** the DNS QNAME (if present) is used as a secondary key for matching queries with responses.

.. _max-transactions:

max-transactions
^^^^^^^^^^^^^^^^

:data node: ``<instance-id>/transaction-table/max-transactions``
:default: 1048576

The value of this parameter must be a power of 2. It specifies the maximum number of pending DNS transactions that DNS Probe can handle at any given time, which in turn affects the size of in-memory transaction table.


src/config/Config.h
/*
 *  Copyright (C) 2018 Brno University of Technology
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *  In addition, as a special exception, the copyright holders give
 *  permission to link the code of portions of this program with the
 *  OpenSSL library under certain conditions as described in each
 *  individual source file, and distribute linked combinations including
 *  the two.
 */

#pragma once


#include <string>
#include <forward_list>
#include <functional>

#include "ConfigTypes.h"
#include "ConfigItem.h"

namespace DDP {
    /**
     * Structure containing configuration for connecting to Kafka cluster
     */
    struct KafkaConfig
    {
        KafkaConfig() : brokers("127.0.0.1"),
                        address_family(KafkaAddressFamily::ANY),
                        topic("dns-probe"),
                        partition(""),
                        ca_location(""),
                        sec_protocol(KafkaSecurityProtocol::PLAINTEXT),
                        cert_location(""),
                        key_location(""),
                        key_passwd(""),
                        sasl_mechanism(KafkaSaslMechanism::PLAIN),
                        sasl_username(""),
                        sasl_password("") {}

        ConfigItem<std::string> brokers; //!< Comma separated list of Kafka brokers (host or host:port)
        ConfigItem<KafkaAddressFamily> address_family; //!< Brokers' address family (any/ipv4/ipv6)
        ConfigItem<std::string> topic; //!< Kafka topic
        ConfigItem<std::string> partition; //!< Kafka message key to assign messages to specific partition

        ConfigItem<std::string> ca_location; //!< CA certificate for authentication of Kafka brokers's certificate
        ConfigItem<KafkaSecurityProtocol> sec_protocol; //!< Protocol used to communicate with Kafka brokesrs
        ConfigItem<std::string> cert_location; //!< Public key for authentication to Kafka cluster
        ConfigItem<std::string> key_location; //!< Private key for authentication to Kafka cluster
        ConfigItem<std::string> key_passwd; //!< Private key passphrase
        ConfigItem<KafkaSaslMechanism> sasl_mechanism; //!< SASL mechanism to use for authentication to Kafka brokers
        ConfigItem<std::string> sasl_username; //!< SASL username
        ConfigItem<std::string> sasl_password; //!< SASL password
    };

    /**
     * Structure containing configuration of the application
     */
    struct Config
    {
        Config() : interface_list(),
                   pcap_list(),
                   raw_pcap(false),
                   dnstap_socket_list(),
                   dnstap_socket_group(),
                   knot_socket_path("/tmp"),
                   knot_socket_count(0),
                   log_file(),
                   coremask(0x7),
                   dns_ports({53}),
                   ipv4_allowlist(),
                   ipv4_denylist(),
                   ipv6_allowlist(),
                   ipv6_denylist(),
                   tt_size(1048576),
                   tt_timeout(1000),
                   match_qname(false),
                   tcp_ct_size(131072),
                   tcp_ct_timeout(60000),
                   target_directory("."),
                   file_prefix("dns_"),
                   file_rot_timeout(0),
                   file_rot_size(0),
                   file_compression(true),
                   pcap_export(PcapExportCfg::DISABLED),
                   country_db(),
                   asn_db(),
                   export_format(ExportFormat::PARQUET),
                   parquet_records(5000000),
                   cdns_fields(get_cdns_bitmask()),
                   cdns_records_per_block(10000),
                   cdns_blocks_per_file(0),
                   cdns_export_resp_rr(false),
                   export_location(ExportLocation::LOCAL),
                   export_ip("127.0.0.1"),
                   export_port(6378),
                   export_ca_cert(),
                   backup_export_ip(""),
                   backup_export_port(6378),
                   kafka_export(),
                   anonymize_ip(false),
                   ip_encryption(IpEncryption::AES),
                   ip_enc_key("key.cryptopant"),
                   export_stats(false),
                   stats_per_ip(false),
                   stats_timeout(300),
                   stats_location(ExportLocation::LOCAL),
                   stats_directory("."),
                   stats_ip("127.0.0.1"),
                   stats_port(6379),
                   stats_ca_cert(),
                   backup_stats_ip(""),
                   backup_stats_port(6379),
                   stats_kafka_export(),
                   moving_avg_window(300),
                   stats_fields(get_stats_bitmask()),
                   instance("default") { stats_kafka_export.topic = ConfigItem<std::string>("dns-probe-stats"); }

        ConfigItem<CList<std::string>> interface_list; //!< List of network interfaces to process traffic from
        ConfigItem<CList<std::string>> pcap_list; //!< List of PCAP files to process
        ConfigItem<bool> raw_pcap; //!< Defines if input PCAP files are without ethernet headers
        ConfigItem<CList<std::string>> dnstap_socket_list; //!< List of unix sockets to process dnstap data from
        ConfigItem<std::string> dnstap_socket_group; //!< User group under which to create dnstap sockets
        ConfigItem<std::string> knot_socket_path; //!< Path to directory in which to create Knot interface sockets
        ConfigItem<uint32_t> knot_socket_count; //!< Number of Knot interface sockets to create
        ConfigItem<std::string> log_file; //!< Log file for storing probe's logs
        ConfigItem<ThreadManager::MaskType> coremask; //!< Coremask used fo selecting cores where application will be running.
        ConfigItem<CList<Port_t>> dns_ports; //!< TCP/UDP port list used for identifying DNS traffic
        ConfigItem<CList<IPv4_prefix_t>> ipv4_allowlist; //!< List of allowed IPv4 addresses to process traffic from
        ConfigItem<CList<IPv4_prefix_t>> ipv4_denylist; //!< List of IPv4 addresses from which to NOT process traffic
        ConfigItem<CList<IPv6_prefix_t>> ipv6_allowlist; //!< List of allowed IPv6 addresses to process traffic from
        ConfigItem<CList<IPv6_prefix_t>> ipv6_denylist; //!< List of IPv6 addresses from which to NOT process traffic

        ConfigItem<uint32_t> tt_size; //!< Number of items in the transaction table
        ConfigItem<uint64_t> tt_timeout; //!< Timeout for orphaned items transaction table in milliseconds
        ConfigItem<bool> match_qname; //!< Enable matching qnames in transaction table

        ConfigItem<uint32_t> tcp_ct_size; //!< Maximal concurrent tracking TCP connections
        ConfigItem<uint64_t> tcp_ct_timeout; //!< Timeout of TCP connection

        ConfigItem<std::string> target_directory; //!< Directory for exported data
        ConfigItem<std::string> file_prefix; //!< Exported file prefix name
        ConfigItem<uint32_t> file_rot_timeout; //!< Exported file rotation timeout in seconds
        ConfigItem<uint64_t> file_rot_size; //!< Exported file size limit in MB
        ConfigItem<bool> file_compression; //!< Enable GZIP compression for exported files
        ConfigItem<PcapExportCfg> pcap_export; //!< Define what will be in exported PCAPs
        ConfigItem<std::string> country_db; //!< Path to Maxmind Country database
        ConfigItem<std::string> asn_db; //!< Path to Maxmind ASN database

        ConfigItem<ExportFormat> export_format; //!< Specify export format
        ConfigItem<uint64_t> parquet_records; //!< Number of records in parquet file
        ConfigBitfield<CdnsBits> cdns_fields; //!< Fields which will be part of CDNS file
        ConfigItem<uint64_t> cdns_records_per_block; //!< Number of records in one block in CDNS file
        ConfigItem<uint64_t> cdns_blocks_per_file; //!< Number of blocks in CDNS file
        ConfigItem<bool> cdns_export_resp_rr; //!< Export full answer and additional RRs of response

        ConfigItem<ExportLocation> export_location; //!< Location for the exported DNS records
        ConfigItem<std::string> export_ip; //!< IP address for remote export of DNS records
        ConfigItem<uint16_t> export_port; //!< Transport protocol port for remote export of DNS records
        ConfigItem<std::string> export_ca_cert; //!< CA certificate for authentication of remote server's certificate
        ConfigItem<std::string> backup_export_ip; //!< Backup IP address for remote export of DNS records
        ConfigItem<uint16_t> backup_export_port; //!< Backup transport protocol port for remote export of DNS records
        KafkaConfig kafka_export; //!< Kafka configuration for export of DNS records

        ConfigItem<bool> anonymize_ip; //!< Enable client IP anonymization in exported data
        ConfigItem<IpEncryption> ip_encryption; //!< Encryption algorithm for IP anonymization
        ConfigItem<std::string> ip_enc_key; //!< File with encryption key for IP anonymization

        ConfigItem<bool> export_stats; //!< Enable export of run-time statistics
        ConfigItem<bool> stats_per_ip; //!< Enable export of 'queries*' run-time statistics per IP address
        ConfigItem<uint32_t> stats_timeout; //!< Export run-time statistics every 'stats_timeout' seconds
        ConfigItem<ExportLocation> stats_location; //!< Location for exported run-time statistics
        ConfigItem<std::string> stats_directory; //!< Directory for exported run-time statistics
        ConfigItem<std::string> stats_ip; //!< IP address for remote export of run-time statistics
        ConfigItem<uint16_t> stats_port; //!< Transport protocol port for remote export of run-time statistics
        ConfigItem<std::string> stats_ca_cert; //!< CA certificate for authentication of remote server's certificate
        ConfigItem<std::string> backup_stats_ip; //!< Backup IP address for remote export of run-time statistics
        ConfigItem<uint16_t> backup_stats_port; //!< Backup transport protocol port for remote export of run-time statistics
        KafkaConfig stats_kafka_export; //!< Kafka configuration for export of run-time statistics
        ConfigItem<uint16_t> moving_avg_window; //!< Time window for computing queries-per-second* statistics
        ConfigBitfield<StatsBits> stats_fields; //!< Indicates which statistics should be exported

        ConfigItem<std::string> instance; //!< Name of running dns-probe instance. "Default" by default.
    };
}
